Preserve the .debug_gdb_scripts section

sebastianpoeplau opened the pull request:
Make sure that compiler and linker don't optimize the section's contents
away by adding the global holding the data to `llvm.used`. This
eliminates the need for a volatile load in the main shim; since the LLVM
codegen backend is the only implementer of the corresponding trait
function, remove it entirely.

Pretty printers in dylib dependencies are now emitted by the main crate
instead of the dylib; apart from matching how rlibs are handled, this
approach has the advantage that `omit_gdb_pretty_printer_section` keeps
working with dylib dependencies.

r? @bjorn3

bjorn3 commented on compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs:
```diff
@@ -80,6 +72,8 @@ pub(crate) fn get_or_insert_gdb_debug_scripts_section_global<'ll>(
             // the string it contains. Otherwise we get a warning from GDB.
             llvm::LLVMSetAlignment(section_var, 1);
+            // Make sure that the linker doesn't optimize the global away.
+            llvm::LLVMRustAppendToUsed(cx.llmod, section_var);
```
This can probably use `cx.add_used_global(section_var)` instead.

bjorn3 commented on compiler/rustc_codegen_gcc/src/debuginfo.rs:
```diff
@@ -254,7 +250,7 @@ impl<'gcc, 'tcx> DebugInfoCodegenMethods<'tcx> for CodegenCx<'gcc, 'tcx> {
     }
 
-    fn debuginfo_finalize(&self) {
+    fn debuginfo_finalize(&mut self) {
```
Can you add a TODO about adding the gdb debug scripts section?

bjorn3 commented:
@bors r+

tgross35 commented:
Looks like some GDB tests failed https://github.com/rust-lang/rust/pull/143755#issuecomment-3059647213
@bors r-
@bors2 try jobs=dist-i586-gnu-i586-i686-musl

sebastianpoeplau commented:
Hmm, the [full log](https://productionresultssa18.blob.core.windows.net/actions-results/af8f077a-6805-4c3b-9bf4-b4db7c183c68/workflow-job-run-d71e1347-eda0-531f-bcb2-eff462d4e37b/logs/job/job-logs.txt?rsct=text%2Fplain&se=2025-07-11T08%3A02%3A58Z&sig=IfAdvPpqQXqFuKcnQTwet9zxu%2Ff2FGcjfxpYx60oXzw%3D&ske=2025-07-11T18%3A54%3A36Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2025-07-11T06%3A54%3A36Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-05-05&sp=r&spr=https&sr=b&st=2025-07-11T07%3A52%3A53Z&sv=2025-05-05) of the CI job led me to believe that it's `i586-unknown-linux-gnu`, but locally the test passes for that target (running `./x.py test tests/debuginfo/embedded-visualizer.rs --target i586-unknown-linux-gnu`). Will try with the musl target :+1: 

I noticed though that my GDB is version 15, whereas the test log shows GDB 12; maybe that's the reason for the discrepancy. In general, the `.debug_gdb_scripts` section looks correct to me...

sebastianpoeplau commented:
Getting closer :yum: I can reproduce the failure in the Docker container via `src/ci/docker/run.sh --dev dist-i586-gnu-i586-i686-musl`, then `../x test tests/debuginfo/embedded-visualizer.rs --target i586-unknown-linux-gnu` inside the container (thanks for the hint!). It seems that the binary just doesn't have a `.debug_gdb_scripts` section :flushed: And even more strangely, the test succeeds with `--target i686-unknown-linux-gnu` (i.e., just replacing i586 with i686). Maybe LLVM somehow doesn't support `llvm.used` on i586? :thinking:

sebastianpoeplau commented:
i686 does indeed use a different linker. The contents of the `.comment` section suggest that `i686-unknown-linux-gnu` uses the system's GCC 11, whereas `i586-unknown-linux-gnu` links with GCC 8 from crosstool-ng. The former comes with a linker that supports `SHF_GNU_RETAIN`, the latter doesn't.

So where does that leave us regarding the test failure on i586? Is it possible to upgrade the crosstool-ng toolchain? Is compatibility with the old ld version required? Can the test simply be disabled for that target?

matthiaskrgr commented:
@bors rollup=iffy
this may bounce a couple more times

rust-log-analyzer commented:
The job **`aarch64-gnu-debug`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16256024130/job/45892508011) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45892508011) [(plain)](https://github.com/rust-lang/rust/commit/f7d013edefafee39f68b74750bb34dad4a23d009/checks/45892508011/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [run-make] tests/run-make/symbols-all-mangled stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/symbols-all-mangled/rmake_out" && env -u RUSTFLAGS AR="ar" BUILD_ROOT="/checkout/obj/build/aarch64-unknown-linux-gnu" CARGO="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="clang" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=aarch64-unknown-linux-gnu" CLANG="/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/bin/clang" CXX="clang++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=aarch64-unknown-linux-gnu" HOST_RUSTC_DYLIB_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/bootstrap-tools/aarch64-unknown-linux-gnu/release/deps:/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/lib/rustlib/aarch64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard cgdata codegen codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontendatomic frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser telemetry textapi textapibinaryreader transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/build/bin/FileCheck" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_LINKER="clang" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="aarch64-unknown-linux-gnu" TARGET_EXE_DYLIB_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib" __RUSTC_DEBUG_ASSERTIONS_ENABLED="1" __STD_DEBUG_ASSERTIONS_ENABLED="1" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/symbols-all-mangled/rmake"
stdout: none
--- stderr -------------------------------

thread 'main' panicked at /checkout/tests/run-make/symbols-all-mangled/rmake.rs:48:9:
Unmangled symbol found in liba_lib.a: __rustc_debug_gdb_scripts_section_std_d85e11bee23a1dc7
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
------------------------------------------



```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #143919) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 commented:
Tests get compiled with `-Cprefer-dynamic` by default to dynamically link against libstd. Try adding `//@ no-prefer-dynamic` to the relevant tests.

@bors r-

sebastianpoeplau commented:
> Try adding //@ no-prefer-dynamic to the relevant tests.

That doesn't seem to help - we get the section from the libstd rlib as well. It seems to be part of all object files inside the archive. I wonder why...

Oh, I think it's because of this bit in `library/std/Cargo.toml`:

```
[lib]
crate-type = ["dylib", "rlib"]
```

The code in `needs_gdb_debug_scripts_section` checks whether _any_ of the crate types needs the section! That doesn't seem right, but reading the comment in that function made me wonder about the approach more generally:

>    // To ensure the section `__rustc_debug_gdb_scripts_section__` will not create
>    // ODR violations at link time, this section will not be emitted for rlibs since
>    // each rlib could produce a different set of visualizers that would be embedded
>    // in the `.debug_gdb_scripts` section. For that reason, we make sure that the
>    // section is only emitted for leaf crates.

This doesn't apply anymore, does it? There's no risk of ODR violations because the symbols get a unique name. The real reason now is that `get_or_insert_gdb_debug_scripts_section_global` still collects visualizers transitively from all crates. If we keep the transitive collection, shouldn't we drop the emission of the debug script section for dylibs? I'm thinking that they shouldn't have the section because the compiler can see their visualizers when building executables that use them, can't it? I've tested locally that returning false for `CrateType::Dylib` fixes the debuginfo tests, but I don't know whether it has unwanted side effects.

Alternatively, we could collect visualizers for the current crate only, emit the per-crate list for all crate types, and rely on the linker to merge the `.debug_gdb_scripts` sections. This would require special handling of `gdb_load_rust_pretty_printers.py` though - maybe it could be emitted only by libstd?

sebastianpoeplau commented:
> the compiler can see their visualizers when building executables that use them

This seems to be incorrect when the dependency is in a dylib crate. I tried a variant of `tests/debuginfo/embedded-visualizer.rs` where the dependency is built as a dylib instead of an rlib. When I don't emit the `.debug_gdb_scripts` section for dylibs, the test fails because the dependency's visualizer is missing. It works fine, however, with an rlib, even if there's no `.debug_gdb_scripts` section in it.

Does this mean that the compiler has more visibility into rlib dependencies than into dylibs?

sebastianpoeplau commented:
Ah, here we go: `collect_debugger_visualizers_transitive` [filters by metadata source](https://github.com/rust-lang/rust/blob/bc1ff10a0f9f01a8548f6a3eb77ed8a43192a5ef/compiler/rustc_codegen_ssa/src/base.rs#L615), apparently excluding dylibs. If I remove the filter and stop the compiler from emitting `.debug_gdb_scripts` for dylibs, all debuginfo tests pass :partying_face:

sebastianpoeplau commented:
Is there a rustbot or bors command that I could use to trigger another test run?

bors commented:
:umbrella: The latest upstream changes (presumably #144249) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 commented:
Bit unfortunate to lose the section for rust dylibs, but given that you would rarely use a rust dylib without an executable, cdylib or staticlib depending on it, it should be fine. The only case I can imagine this reducing debugability is when you use a rust dylib as plugin loaded with dlopen and this rust dylib uses `#![debugger_visualizer(gdb_script_file = "...")]`. That can be fixed later by replacing `#![omit_gdb_pretty_printer_section]` usage in the test suite with telling gdb to ignore `.debug_gdb_scripts` I think. The attribute is unstable anyway and only meant for usage inside the test suite.

@bors r+

bors commented:
:hourglass: Testing commit 2133b4fffbc222e1145e2f2592c545187bbcadd9 with merge 09737a649250b08897c8b0152d72ec4334dc6d4a...
<!-- homu: {"type":"BuildStarted","head_sha":"2133b4fffbc222e1145e2f2592c545187bbcadd9","merge_sha":"09737a649250b08897c8b0152d72ec4334dc6d4a"} -->

rust-log-analyzer commented:
The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16543667225/job/46788382117) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46788382117) [(plain)](https://github.com/rust-lang/rust/commit/09737a649250b08897c8b0152d72ec4334dc6d4a/checks/46788382117/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [debuginfo-gdb] tests/debuginfo/empty-string.rs stdout ----
NOTE: compiletest thinks it is using GDB version 12001000

error: check directive(s) from `/checkout/tests/debuginfo/empty-string.rs` not found in debugger output. errors:
    (empty-string.rs:10) `$1 = ""`
the following subset of check directive(s) was found successfully:
    (empty-string.rs:13) `$2 = ""`
status: exit status: 0
command: PYTHONPATH="/checkout/src/etc" "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/empty-string.gdb/empty-string.debugger.script"
--- stdout -------------------------------
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x14c8: file /checkout/tests/debuginfo/empty-string.rs, line 30.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, empty_string::main () at /checkout/tests/debuginfo/empty-string.rs:30
30     zzz(); // #break
$1 = alloc::string::String {vec: alloc::vec::Vec<u8, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<u8, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x1}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (0), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<u8>}, len: 0}}
$2 = ""
A debugging session is active.

 Inferior 1 [process 471737] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr: none


---- [debuginfo-gdb] tests/debuginfo/pretty-huge-vec.rs stdout ----
NOTE: compiletest thinks it is using GDB version 12001000

error: check directive(s) from `/checkout/tests/debuginfo/pretty-huge-vec.rs` not found in debugger output. errors:
    (pretty-huge-vec.rs:11) `$1 = Vec(size=1000000000) = {[...]...}`
    (pretty-huge-vec.rs:14) `$2 = &[u8](size=1000000000) = {[...]...}`
status: exit status: 0
command: PYTHONPATH="/checkout/src/etc" "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-huge-vec.gdb/pretty-huge-vec.debugger.script"
--- stdout -------------------------------
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x1ba7: file /checkout/tests/debuginfo/pretty-huge-vec.rs, line 25.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_huge_vec::main () at /checkout/tests/debuginfo/pretty-huge-vec.rs:25
25     zzz(); // #break
$1 = alloc::vec::Vec<u8, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<u8, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0xbc271010}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (1000000000), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<u8>}, len: 1000000000}
$2 = &[u8] {data_ptr: 0xbc271010, length: 1000000000}
A debugging session is active.

 Inferior 1 [process 472785] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr: none


---- [debuginfo-gdb] tests/debuginfo/pretty-slices.rs stdout ----
NOTE: compiletest thinks it is using GDB version 12001000

error: check directive(s) from `/checkout/tests/debuginfo/pretty-slices.rs` not found in debugger output. errors:
    (pretty-slices.rs:8) ` $1 = &[i32](size=3) = {0, 1, 2}`
    (pretty-slices.rs:11) ` $2 = &mut [i32](size=4) = {2, 3, 5, 7}`
    (pretty-slices.rs:17) ` $4 = "mutable string slice"`
the following subset of check directive(s) was found successfully:
    (pretty-slices.rs:14) `$3 = "string slice"`
status: exit status: 0
command: PYTHONPATH="/checkout/src/etc" "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-slices.gdb/pretty-slices.debugger.script"
--- stdout -------------------------------
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x21b8: file /checkout/tests/debuginfo/pretty-slices.rs, line 43.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_slices::main () at /checkout/tests/debuginfo/pretty-slices.rs:43
43     b(); // #break
$1 = &[i32] {data_ptr: 0x56558544, length: 3}
$2 = &mut [i32] {data_ptr: 0xffffcf64, length: 4}
$3 = "string slice"
$4 = &mut str {data_ptr: 0x5655b540, length: 20}
A debugging session is active.

 Inferior 1 [process 472803] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr: none


---- [debuginfo-gdb] tests/debuginfo/pretty-std.rs stdout ----
NOTE: compiletest thinks it is using GDB version 12001000

error: check directive(s) from `/checkout/tests/debuginfo/pretty-std.rs` not found in debugger output. errors:
    (pretty-std.rs:13) `$1 = &[i32](size=4) = {0, 1, 2, 3}`
    (pretty-std.rs:16) `$2 = Vec(size=4) = {4, 5, 6, 7}`
    (pretty-std.rs:22) `$4 = "IAMA string!"`
    (pretty-std.rs:31) `$7 = "IAMA OS string 😃"`
    (pretty-std.rs:34) `$8 = core::option::Option<alloc::string::String>::Some("IAMA optional string!")`
    (pretty-std.rs:38) `$9 = core::option::Option<alloc::string::String>::Some("IAMA "...)`
the following subset of check directive(s) was found successfully:
    (pretty-std.rs:19) `$3 = "IAMA string slice!"`
    (pretty-std.rs:25) `$5 = core::option::Option<i16>::Some(8)`
    (pretty-std.rs:28) `$6 = core::option::Option<i64>::None`
status: exit status: 0
command: PYTHONPATH="/checkout/src/etc" "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-std.gdb/pretty-std.debugger.script"
--- stdout -------------------------------
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x48da: file /checkout/tests/debuginfo/pretty-std.rs, line 181.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_std::main () at /checkout/tests/debuginfo/pretty-std.rs:181
181     zzz(); // #break
$1 = &[i32] {data_ptr: 0x5655a920, length: 4}
$2 = alloc::vec::Vec<u64, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<u64, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x5655e420}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (4), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<u64>}, len: 4}
$3 = "IAMA string slice!"
$4 = alloc::string::String {vec: alloc::vec::Vec<u8, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<u8, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x5655e540}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (12), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<u8>}, len: 12}}
$5 = core::option::Option<i16>::Some(8)
$6 = core::option::Option<i64>::None
$7 = std::ffi::os_str::OsString {inner: std::sys::os_str::bytes::Buf {inner: alloc::vec::Vec<u8, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<u8, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x5655e550}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (19), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<u8>}, len: 19}}}
$8 = core::option::Option<alloc::string::String>::Some(alloc::string::String {vec: alloc::vec::Vec<u8, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<u8, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x5655e570}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (21), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<u8>}, len: 21}})
$9 = core::option::Option<alloc::string::String>::Some(alloc::string::String {vec: alloc::vec::Vec<u8, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<u8, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x5655e570}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (21), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<u8>}, len: 21}})
A debugging session is active.

 Inferior 1 [process 472853] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr: none


---- [debuginfo-gdb] tests/debuginfo/pretty-uninitialized-vec.rs stdout ----
NOTE: compiletest thinks it is using GDB version 12001000

error: check directive(s) from `/checkout/tests/debuginfo/pretty-uninitialized-vec.rs` not found in debugger output. errors:
    (pretty-uninitialized-vec.rs:10) `$1 = Vec(size=[...])[...]`
status: exit status: 0
command: PYTHONPATH="/checkout/src/etc" "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-uninitialized-vec.gdb/pretty-uninitialized-vec.debugger.script"
--- stdout -------------------------------
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x1bc4: file /checkout/tests/debuginfo/pretty-uninitialized-vec.rs, line 18.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_uninitialized_vec::main () at /checkout/tests/debuginfo/pretty-uninitialized-vec.rs:18
18     zzz(); // #break
$1 = alloc::vec::Vec<i32, alloc::alloc::Global> {buf: alloc::raw_vec::RawVec<i32, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0xffffff60}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (4159000576), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<i32>}, len: 0}
A debugging session is active.

 Inferior 1 [process 472857] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr: none


---- [debuginfo-gdb] tests/debuginfo/rc_arc.rs stdout ----
NOTE: compiletest thinks it is using GDB version 12001000

error: check directive(s) from `/checkout/tests/debuginfo/rc_arc.rs` not found in debugger output. errors:
    (rc_arc.rs:11) `[...]$1 = Rc(strong=11, weak=1) = {value = 111, strong = 11, weak = 1}`
    (rc_arc.rs:13) `[...]$2 = Arc(strong=21, weak=1) = {value = 222, strong = 21, weak = 1}`
status: exit status: 0
command: PYTHONPATH="/checkout/src/etc" "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/rc_arc.gdb/rc_arc.debugger.script"
--- stdout -------------------------------
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x704a: file /checkout/tests/debuginfo/rc_arc.rs, line 133.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, rc_arc::main () at /checkout/tests/debuginfo/rc_arc.rs:133
133     zzz(); // #break
$1 = alloc::rc::Rc<i32, alloc::alloc::Global> {ptr: core::ptr::non_null::NonNull<alloc::rc::RcInner<i32>> {pointer: 0x56562540}, phantom: core::marker::PhantomData<alloc::rc::RcInner<i32>>, alloc: alloc::alloc::Global}
$2 = alloc::sync::Arc<i32, alloc::alloc::Global> {ptr: core::ptr::non_null::NonNull<alloc::sync::ArcInner<i32>> {pointer: 0x56562550}, phantom: core::marker::PhantomData<alloc::sync::ArcInner<i32>>, alloc: alloc::alloc::Global}
A debugging session is active.

 Inferior 1 [process 472912] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr: none


---- [debuginfo-gdb] tests/debuginfo/pretty-std-collections.rs stdout ----
NOTE: compiletest thinks it is using GDB version 12001000

error: check directive(s) from `/checkout/tests/debuginfo/pretty-std-collections.rs` not found in debugger output. errors:
    (pretty-std-collections.rs:10) `$1 = BTreeSet(size=15) = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14}`
    (pretty-std-collections.rs:13) `$2 = BTreeSet(size=0)`
    (pretty-std-collections.rs:16) `$3 = BTreeMap(size=15) = {[0] = 0, [1] = 1, [2] = 2, [3] = 3, [4] = 4, [5] = 5, [6] = 6, [7] = 7, [8] = 8, [9] = 9, [10] = 10, [11] = 11, [12] = 12, [13] = 13, [14] = 14}`
    (pretty-std-collections.rs:19) `$4 = BTreeMap(size=0)`
    (pretty-std-collections.rs:22) `$5 = BTreeMap(size=2) = {[false] = [...], [true] = [...]}`
    (pretty-std-collections.rs:26) `$6 = BTreeMap(size=15) = {[0] = pretty_std_collections::MyLeafNode (0), [...]}`
    (pretty-std-collections.rs:30) `$7 = BTreeMap(size=1) = {[()] = 1}`
    (pretty-std-collections.rs:33) `$8 = BTreeMap(size=1) = {[1] = ()}`
    (pretty-std-collections.rs:36) `$9 = BTreeMap(size=1) = {[()] = ()}`
    (pretty-std-collections.rs:39) `$10 = VecDeque(size=3) = {5, 3, 7}`
    (pretty-std-collections.rs:42) `$11 = VecDeque(size=7) = {2, 3, 4, 5, 6, 7, 8}`
    (pretty-std-collections.rs:45) `$12 = HashMap(size=4) = {[1] = 10, [2] = 20, [3] = 30, [4] = 40}`
    (pretty-std-collections.rs:48) `$13 = HashSet(size=4) = {1, 2, 3, 4}`
status: exit status: 0
command: PYTHONPATH="/checkout/src/etc" "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-std-collections.gdb/pretty-std-collections.debugger.script"
--- stdout -------------------------------
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x3757b: file /checkout/tests/debuginfo/pretty-std-collections.rs, line 146.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_std_collections::main () at /checkout/tests/debuginfo/pretty-std-collections.rs:146
146     zzz(); // #break
$1 = alloc::collections::btree::set::BTreeSet<i32, alloc::alloc::Global> {map: alloc::collections::btree::map::BTreeMap<i32, alloc::collections::btree::set_val::SetValZST, alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, alloc::collections::btree::set_val::SetValZST, alloc::collections::btree::node::marker::LeafOrInternal>>::Some(alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, alloc::collections::btree::set_val::SetValZST, alloc::collections::btree::node::marker::LeafOrInternal> {height: 1, node: core::ptr::non_null::NonNull<alloc::collections::btree::node::LeafNode<i32, alloc::collections::btree::set_val::SetValZST>> {pointer: 0x5659f5c0}, _marker: core::marker::PhantomData<(alloc::collections::btree::node::marker::Owned, alloc::collections::btree::node::marker::LeafOrInternal)>}), length: 15, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<(i32, alloc::collections::btree::set_val::SetValZST), alloc::alloc::Global>>}}
$2 = alloc::collections::btree::set::BTreeSet<i32, alloc::alloc::Global> {map: alloc::collections::btree::map::BTreeMap<i32, alloc::collections::btree::set_val::SetValZST, alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, alloc::collections::btree::set_val::SetValZST, alloc::collections::btree::node::marker::LeafOrInternal>>::None, length: 0, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<(i32, alloc::collections::btree::set_val::SetValZST), alloc::alloc::Global>>}}
$3 = alloc::collections::btree::map::BTreeMap<i32, i32, alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, i32, alloc::collections::btree::node::marker::LeafOrInternal>>::Some(alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, i32, alloc::collections::btree::node::marker::LeafOrInternal> {height: 1, node: core::ptr::non_null::NonNull<alloc::collections::btree::node::LeafNode<i32, i32>> {pointer: 0x5659f870}, _marker: core::marker::PhantomData<(alloc::collections::btree::node::marker::Owned, alloc::collections::btree::node::marker::LeafOrInternal)>}), length: 15, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<(i32, i32), alloc::alloc::Global>>}
$4 = alloc::collections::btree::map::BTreeMap<i32, u32, alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, u32, alloc::collections::btree::node::marker::LeafOrInternal>>::None, length: 0, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<(i32, u32), alloc::alloc::Global>>}
$5 = alloc::collections::btree::map::BTreeMap<bool, core::option::Option<bool>, alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, bool, core::option::Option<bool>, alloc::collections::btree::node::marker::LeafOrInternal>>::Some(alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, bool, core::option::Option<bool>, alloc::collections::btree::node::marker::LeafOrInternal> {height: 0, node: core::ptr::non_null::NonNull<alloc::collections::btree::node::LeafNode<bool, core::option::Option<bool>>> {pointer: 0x5659f420}, _marker: core::marker::PhantomData<(alloc::collections::btree::node::marker::Owned, alloc::collections::btree::node::marker::LeafOrInternal)>}), length: 2, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<(bool, core::option::Option<bool>), alloc::alloc::Global>>}
$6 = alloc::collections::btree::map::BTreeMap<i32, pretty_std_collections::MyLeafNode, alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, pretty_std_collections::MyLeafNode, alloc::collections::btree::node::marker::LeafOrInternal>>::Some(alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, pretty_std_collections::MyLeafNode, alloc::collections::btree::node::marker::LeafOrInternal> {height: 1, node: core::ptr::non_null::NonNull<alloc::collections::btree::node::LeafNode<i32, pretty_std_collections::MyLeafNode>> {pointer: 0x5659f9f0}, _marker: core::marker::PhantomData<(alloc::collections::btree::node::marker::Owned, alloc::collections::btree::node::marker::LeafOrInternal)>}), length: 15, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<(i32, pretty_std_collections::MyLeafNode), alloc::alloc::Global>>}
$7 = alloc::collections::btree::map::BTreeMap<(), i32, alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, (), i32, alloc::collections::btree::node::marker::LeafOrInternal>>::Some(alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, (), i32, alloc::collections::btree::node::marker::LeafOrInternal> {height: 0, node: core::ptr::non_null::NonNull<alloc::collections::btree::node::LeafNode<(), i32>> {pointer: 0x5659f710}, _marker: core::marker::PhantomData<(alloc::collections::btree::node::marker::Owned, alloc::collections::btree::node::marker::LeafOrInternal)>}), length: 1, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<((), i32), alloc::alloc::Global>>}
$8 = alloc::collections::btree::map::BTreeMap<i32, (), alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, (), alloc::collections::btree::node::marker::LeafOrInternal>>::Some(alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, i32, (), alloc::collections::btree::node::marker::LeafOrInternal> {height: 0, node: core::ptr::non_null::NonNull<alloc::collections::btree::node::LeafNode<i32, ()>> {pointer: 0x5659fa90}, _marker: core::marker::PhantomData<(alloc::collections::btree::node::marker::Owned, alloc::collections::btree::node::marker::LeafOrInternal)>}), length: 1, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<(i32, ()), alloc::alloc::Global>>}
$9 = alloc::collections::btree::map::BTreeMap<(), (), alloc::alloc::Global> {root: core::option::Option<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, (), (), alloc::collections::btree::node::marker::LeafOrInternal>>::Some(alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, (), (), alloc::collections::btree::node::marker::LeafOrInternal> {height: 0, node: core::ptr::non_null::NonNull<alloc::collections::btree::node::LeafNode<(), ()>> {pointer: 0x5659f750}, _marker: core::marker::PhantomData<(alloc::collections::btree::node::marker::Owned, alloc::collections::btree::node::marker::LeafOrInternal)>}), length: 1, alloc: core::mem::manually_drop::ManuallyDrop<alloc::alloc::Global> {value: alloc::alloc::Global}, _marker: core::marker::PhantomData<alloc::boxed::Box<((), ()), alloc::alloc::Global>>}
$10 = alloc::collections::vec_deque::VecDeque<i32, alloc::alloc::Global> {head: 0, len: 3, buf: alloc::raw_vec::RawVec<i32, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x5659fad0}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (4), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<i32>}}
$11 = alloc::collections::vec_deque::VecDeque<i32, alloc::alloc::Global> {head: 1, len: 7, buf: alloc::raw_vec::RawVec<i32, alloc::alloc::Global> {inner: alloc::raw_vec::RawVecInner<alloc::alloc::Global> {ptr: core::ptr::unique::Unique<u8> {pointer: core::ptr::non_null::NonNull<u8> {pointer: 0x5659faf0}, _marker: core::marker::PhantomData<u8>}, cap: core::num::niche_types::UsizeNoHighBit (8), alloc: alloc::alloc::Global}, _marker: core::marker::PhantomData<i32>}}
$12 = std::collections::hash::map::HashMap<u64, u64, core::hash::BuildHasherDefault<pretty_std_collections::SimpleHasher>> {base: hashbrown::map::HashMap<u64, u64, core::hash::BuildHasherDefault<pretty_std_collections::SimpleHasher>, alloc::alloc::Global> {hash_builder: core::hash::BuildHasherDefault<pretty_std_collections::SimpleHasher> (core::marker::PhantomData<fn() -> pretty_std_collections::SimpleHasher>), table: hashbrown::raw::RawTable<(u64, u64), alloc::alloc::Global> {table: hashbrown::raw::RawTableInner {bucket_mask: 7, ctrl: core::ptr::non_null::NonNull<u8> {pointer: 0x5659f410}, growth_left: 3, items: 4}, alloc: alloc::alloc::Global, marker: core::marker::PhantomData<(u64, u64)>}}}
$13 = std::collections::hash::set::HashSet<u64, core::hash::BuildHasherDefault<pretty_std_collections::SimpleHasher>> {base: hashbrown::set::HashSet<u64, core::hash::BuildHasherDefault<pretty_std_collections::SimpleHasher>, alloc::alloc::Global> {map: hashbrown::map::HashMap<u64, (), core::hash::BuildHasherDefault<pretty_std_collections::SimpleHasher>, alloc::alloc::Global> {hash_builder: core::hash::BuildHasherDefault<pretty_std_collections::SimpleHasher> (core::marker::PhantomData<fn() -> pretty_std_collections::SimpleHasher>), table: hashbrown::raw::RawTable<(u64, ()), alloc::alloc::Global> {table: hashbrown::raw::RawTableInner {bucket_mask: 7, ctrl: core::ptr::non_null::NonNull<u8> {pointer: 0x5659fb60}, growth_left: 3, items: 4}, alloc: alloc::alloc::Global, marker: core::marker::PhantomData<(u64, ())>}}}}
A debugging session is active.

 Inferior 1 [process 472992] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr: none



```

</details>

bjorn3 commented:
@bors r+

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16572974744/job/46876309953)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/16572974744/job/46876309953","builder_name":"checks-actions"} -->

bjorn3 commented:
> That can be fixed later by replacing #![omit_gdb_pretty_printer_section] usage in the test suite with telling gdb to ignore .debug_gdb_scripts I think.

Opened https://github.com/rust-lang/rust/pull/144738 for this.

sebastianpoeplau committed:
Preserve the .debug_gdb_scripts section

Make sure that compiler and linker don't optimize the section's contents
away by adding the global holding the data to "llvm.used". The volatile
load in the main shim is retained because "llvm.used", which translates
to SHF_GNU_RETAIN on ELF targets, requires a reasonably recent linker;
emitting the volatile load ensures compatibility with older linkers, at
least when libstd is used.

Pretty printers in dylib dependencies are now emitted by the main crate
instead of the dylib; apart from matching how rlibs are handled, this
approach has the advantage that `omit_gdb_pretty_printer_section` keeps
working with dylib dependencies.

sebastianpoeplau committed:
Embed GDB pretty printers in rlibs and dylibs

Instead of collecting pretty printers transitively when building
executables/staticlibs/cdylibs, let the debugger find each crate's
pretty printers via its .debug_gdb_scripts section. This covers the case
where libraries defining custom pretty printers are loaded dynamically.

sebastianpoeplau commented:
I've changed the code to add the global to `llvm.used` _in addition to_ the volatile load. The second commit drops the transitive collection of pretty printers and therefore depends on your fix for `omit_gdb_pretty_printer_section` in #144738.

bjorn3 commented on compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs:
```diff
@@ -96,31 +100,14 @@ pub(crate) fn needs_gdb_debug_scripts_section(cx: &CodegenCx<'_, '_>) -> bool {
+    // seem like a good default, since this is a rare use case and we don't
+    // want to slow down the common case.
+    let proc_macro_only =
+        cx.tcx.crate_types().iter().all(|&crate_type| crate_type == CrateType::ProcMacro);
```
I think removing this exception would make sense. Proc macros are generally compiled with debuginfo disabled nowadays, so performance isn't really a reason not to include the section for proc macros anymore.

sebastianpoeplau replied:
Good to know - removed :+1:

bjorn3 commented:
Bors forgot to mention it, but now that https://github.com/rust-lang/rust/pull/144738 has landed there is a minor conflict.

bjorn3 commented on compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs:
```diff
@@ -84,35 +95,5 @@ pub(crate) fn get_or_insert_gdb_debug_scripts_section_global<'ll>(
 }
```
I just noticed that the section is created with LinkOnceODR linkage. This was fine when all global visualizers were merged together and a single section was created across all crates, but now that we create different sections for each crate, that would cause one of the sections to be chosen non-deterministically. I think it would be better to use regular linkage, but instead enable string merging. This is also what the GDB manual page for `.debug_gdb_scripts` suggests. I think the way to do that in LLVM is to put the `\x01gdb_load_rust_pretty_printers.py\0` as well as every visualizer in a separate constant global variables with unnamed_addr but still in the same `.debug_gdb_scripts` section.

bjorn3 replied:
> My understanding based on the LLVM Language Reference is that the attribute affects how the linker treats multiple globals of the same name (choosing one of them non-deterministically), but here we're creating globals with different names, unique per crate.

Right missed that. In that case string merging would just be a binary size optimization that can be left for a future PR.

> What do you mean by "instead enable string merging"? Linkage type appending?

No, most object file formats have a feature where you can mark a section as string merging, which causes the linker to interpret the section as a sequence of C strings and deduplicate identical strings. LLVM should automatically treat constant global variables marked as unnamed_addr as C strings if they end with a NUL byte and don't have any internal NUL bytes.

bjorn3 commented:
@bors r+

bors commented:
:hourglass: Testing commit cc2d4b0b09b99498f8b63ae8f9febd37efc5ce83 with merge 97bb07563e9b3a9e725b37f5576351750e252953...
<!-- homu: {"type":"BuildStarted","head_sha":"cc2d4b0b09b99498f8b63ae8f9febd37efc5ce83","merge_sha":"97bb07563e9b3a9e725b37f5576351750e252953"} -->

sebastianpoeplau commented:
There's a problem with old linkers again: We're emitting the volatile load in the main shim, but pretty printers are now collected per crate, meaning that the load only preserves the main crate's pretty printers. As far as I can see, this affects only a single test (which uses pretty printers from an rlib). Maybe it's okay to disable it, expecting that people generally have more recent linkers than the CI setup?

bjorn3 commented:
@bors r+

bors commented:
:hourglass: Testing commit b4d923cea0509933b1fb859930cb20784251f9be with merge d82cad87289c9d20df43acb0c951c71c56208fc3...
<!-- homu: {"type":"BuildStarted","head_sha":"b4d923cea0509933b1fb859930cb20784251f9be","merge_sha":"d82cad87289c9d20df43acb0c951c71c56208fc3"} -->

Zalathar commented:
Yielding to enclosing rollup:

@bors retry

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16777707851/job/47524399162)
Approved by: bjorn3
Pushing 8fb40f798a23adf608182ce5f4eb151fdc8e0da5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/16777707851/job/47524399162"},"merge_sha":"8fb40f798a23adf608182ce5f4eb151fdc8e0da5"} -->

rust-timer commented:
Finished benchmarking commit (8fb40f798a23adf608182ce5f4eb151fdc8e0da5): [comparison URL](https://perf.rust-lang.org/compare.html?start=dc0bae1db725fbba8524f195f74f680995fd549e&end=8fb40f798a23adf608182ce5f4eb151fdc8e0da5&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=dc0bae1db725fbba8524f195f74f680995fd549e&end=8fb40f798a23adf608182ce5f4eb151fdc8e0da5&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean   | range           | count |
|:----------------------------------:|:------:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 103.7% | [55.2%, 152.2%] | 2     |
| Regressions ❌ <br /> (secondary)  | 0.5%   | [0.5%, 0.6%]    | 3     |
| Improvements ✅ <br /> (primary)   | -      | -               | 0     |
| Improvements ✅ <br /> (secondary) | -1.8%  | [-4.6%, -0.0%]  | 6     |
| All ❌✅ (primary)                 | 103.7% | [55.2%, 152.2%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=dc0bae1db725fbba8524f195f74f680995fd549e&end=8fb40f798a23adf608182ce5f4eb151fdc8e0da5&stat=max-rss)
<details>
<summary>Results (primary 2.6%, secondary 3.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6% | [2.2%, 2.8%] | 3     |
| Regressions ❌ <br /> (secondary)  | 3.0% | [3.0%, 3.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.6% | [2.2%, 2.8%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=dc0bae1db725fbba8524f195f74f680995fd549e&end=8fb40f798a23adf608182ce5f4eb151fdc8e0da5&stat=cycles:u)
<details>
<summary>Results (primary 99.0%, secondary 0.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 99.0% | [53.9%, 144.0%] | 2     |
| Regressions ❌ <br /> (secondary)  | 4.0%  | [2.8%, 6.4%]    | 4     |
| Improvements ✅ <br /> (primary)   | -     | -               | 0     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.8%, -1.4%]  | 4     |
| All ❌✅ (primary)                 | 99.0% | [53.9%, 144.0%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=dc0bae1db725fbba8524f195f74f680995fd549e&end=8fb40f798a23adf608182ce5f4eb151fdc8e0da5&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.2%, secondary 0.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2% | [0.0%, 0.8%] | 46    |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.0%, 1.3%] | 86    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.2% | [0.0%, 0.8%] | 46    |
</details>

**Bootstrap**: 463.057s -> 464.179s (0.24%)
**Artifact size**: 377.38 MiB -> 377.38 MiB (-0.00%)

sebastianpoeplau commented:
Is running rustc-perf locally the way to debug the performance issue? And do you have any hints on how to investigate the cause for the regression, besides checking the `debugger_visualizers` query?

Kobzol commented:
Running rustc-perf locally would probably be a way of debugging this, yeah. https://perf.rust-lang.org/detailed-query.html?commit=8fb40f798a23adf608182ce5f4eb151fdc8e0da5&benchmark=ripgrep-14.1.1-opt&scenario=incr-unchanged&base_commit=dc0bae1db725fbba8524f195f74f680995fd549e seems like some additional CGU was dirtied during incremental. Tbh I would love to have better tooling to debug these incremental invalidations issues, but I think we currently don't have a lot of great tooling for this. It's definitely planned..
